mqtt/mqtt-broker/src/broker.rs

```diff
@@ -1023,26 +1376,66 @@ pub(crate) mod tests {
             Message::Client(_, ClientEvent::ConnAck(proto::ConnAck {
                 return_code:
                     proto::ConnectReturnCode::Refused(
-                        proto::ConnectionRefusedReason::UnacceptableProtocolVersion,
+                        proto::ConnectionRefusedReason::ServerUnavailable,
                     ),
                 ..
             }))
         );
         assert_matches!(
-            rx1.recv().await.unwrap(),
-            Message::Client(_, ClientEvent::DropConnection)
+            rx1.recv().await,
+            Some(Message::Client(_, ClientEvent::DropConnection))
         );
-        assert!(rx1.recv().await.is_none());
+        assert_matches!(rx1.recv().await, None)
+    }
+
+    fn authorize_fn<F>(f: F) -> TestAuthorizer<F>
+    where
+        F: Fn(AuthId) -> Result<bool, Error>,
+    {
+        TestAuthorizer(f)
+    }
+
+    struct TestAuthorizer<F>(F);
+
+    impl<F> Authorizer for TestAuthorizer<F>
+    where
+        F: Fn(AuthId) -> Result<bool, Error>,
+    {
+        fn authorize(&self, auth_id: AuthId) -> Result<bool, Error> {
+            self.0(auth_id)
+        }
+    }
+
+    fn authenticate_fn<F>(f: F) -> TestAuthenticator<F>
+    where
+        F: Fn() -> Result<Option<AuthId>, Error>,
+    {
+        TestAuthenticator(f)
+    }
+
+    struct TestAuthenticator<F>(F);
+
+    impl<F> Authenticator for TestAuthenticator<F>
+    where
+        F: Fn() -> Result<Option<AuthId>, Error>,
+    {
+        fn authenticate(&self, _credentials: Credentials) -> Result<Option<AuthId>, Error> {
+            self.0()
+        }
     }
```

Comment:
You should be able to impl `Authenticator` for functions (in the main code):

```rust
impl<F> Authenticator for F
where
    F: FnOnce(Credentials) -> Result<Option<AuthId>, Error>,
{
    fn authenticate(&self, credentials: Credentials) -> Result<Option<AuthId, Error> {
        self(credentials)
    }
} 
```

then you could pass a closure directly to the builder. (note I didn't compile this code, so it might need to be tweaked)

Reply:
Nice Idea, for tests I can use the same!